runners/overlay-runner/runner.go

```diff
@@ -1,71 +1,95 @@
 package runner
 
 import (
-	"context"
-	"io"
-	"time"
+	"fmt"
+	"os"
 
 	"github.com/tinyci/ci-agents/clients/log"
-	"github.com/tinyci/ci-agents/model"
+	logsvc "github.com/tinyci/ci-agents/clients/log"
+	"github.com/tinyci/ci-agents/clients/queue"
+	"github.com/tinyci/ci-agents/errors"
+	"github.com/tinyci/ci-runners/fw"
+	fwConfig "github.com/tinyci/ci-runners/fw/config"
 	"github.com/tinyci/ci-runners/runners/overlay-runner/config"
 )
 
-// Run is the encapsulation of a CI run.
-type Run struct {
-	Logger      *log.SubLogger
-	QueueItem   *model.QueueItem
-	Config      *config.Config
-	ContainerID string
-	Status      bool
-	Context     context.Context
-	Cancel      context.CancelFunc
+// Runner encapsulates an infinite lifecycle overlay-runner.
+type Runner struct {
+	Config     *config.Config
+	CurrentRun *Run
 }
 
-// NewRun constructs a new *Run.
-func NewRun(context context.Context, cancelFunc context.CancelFunc, qi *model.QueueItem, c *config.Config, logger *log.SubLogger) *Run {
-	if logger == nil {
-		logger = c.Clients.Log
+// Init is the bootstrap of the runner.
+func (r *Runner) Init(ctx *fw.Context) *errors.Error {
+	// we reload the clients on each run
+	r.Config = &config.Config{C: fwConfig.Config{Clients: &fwConfig.Clients{}}}
+	err := fwConfig.Load(ctx.CLIContext.GlobalString("config"), r.Config)
+	if err != nil {
+		return err
 	}
 
-	return &Run{
-		QueueItem: qi,
-		Config:    c,
-		Context:   context,
-		Cancel:    cancelFunc,
-		Logger:    logger,
+	if err := r.Config.Runner.Validate(); err != nil {
+		return err
 	}
-}
 
-// StartCancelFunc launches a goroutine which waits for the cancel signal.
-// Terminates when the run ends; one way or another. This function does not
-// block.
-func (r *Run) StartCancelFunc() {
-	go func() {
-		for {
-			select {
-			case <-r.Context.Done():
-				return
-			default:
-			}
-
-			state, err := r.Config.Clients.Queue.GetCancel(r.QueueItem.Run.ID)
-			if err != nil || !state {
-				time.Sleep(time.Second)
-				continue
-			}
-
-			r.Cancel()
-			return
+	if r.Config.C.Hostname == "" {
+		hostname, err := os.Hostname()
+		if err != nil {
+			return errors.New(err).Wrap("Could not retrieve hostname")
 		}
-	}()
+		r.Config.C.Hostname = hostname
+	}
+
+	r.Config.C.Clients.Log = r.Config.C.Clients.Log.WithFields(log.FieldMap{"hostname": r.Config.C.Hostname})
+
+	return nil
 }
 
-// StartLogger starts a goroutine that writes data produced on the reader to
-// the log.
-func (r *Run) StartLogger(rc io.Reader) {
-	go func() {
-		if err := r.Config.Clients.Asset.Write(r.QueueItem.Run.ID, rc); err != nil {
-			r.Logger.Error(err.Wrapf("Writing log for Run ID %d", r.QueueItem.Run.ID))
-		}
-	}()
+// BeforeRun is executed before the next run is started.
+func (r *Runner) BeforeRun(ctx *fw.Context) *errors.Error {
+	r.CurrentRun = NewRun(ctx.RunCtx, ctx.RunCancelFunc, ctx.QueueItem, r.Config, r.LogsvcClient(ctx))
```

Comment:
This line seems like something that the framework should do automatically. Will each runner need to set up `CurrentRun` in their `BeforeRun`?

Reply:
this is a implementation specific bit; not related to the framework.